lib/google/gax/api_call.rb

```diff
@@ -45,74 +46,63 @@ class ApiCall
-        @params_extractor      = params_extractor
-        @exception_transformer = exception_transformer
+      def initialize stub_method
+        @stub_method = stub_method
       end
 
       ##
       # Invoke the API call.
       #
       # @param request [Object] The request object.
-      # @param options [CallOption, Hash] The options for making the API call.
-      # @param block [Proc] The proc to call when the API call is made.
+      # @param options [ApiCall::Options, Hash] The options for making the API call. A Hash can be provided to customize
+      #   the options object, using keys that match the arguments for {ApiCall::Options.new}. This object should only be
+      #   used once.
+      # @param on_response [Proc] A Proc object to format the response object. The Proc should accept response as an
+      #   argument, and return a formatted response object. Optional.
+      # @param on_operation [Proc] A Proc object to provide a callback of the response and operation objects. The Proc
+      #   will be called with both the response and operation objects. Optional.
+      # @param on_stream [Proc] A Proc object to provide a callback for every streamed response received. The Proc will
+      #   be called with the response object. Should only be used on Bidi and Server streaming RPC calls. Optional.
       #
-      def call request, options: nil, &block
-        options = init_call_options options
-
-        apply_params_extractor! request, options
-
+      # @return [Object, Thread] The response object. Or, when `on_stream` is provided, a thread running the callback
+      #   on streamed response is returned.
+      #
+      def call request, options: nil, on_response: nil, on_operation: nil, on_stream: nil
+        # Converts hash and nil to an options object
+        options = ApiCall::Options.new options.to_h if options.respond_to? :to_h
+        block = compose_stream_proc on_stream: on_stream, on_response: on_response
         deadline = calculate_deadline options
 
         begin
-          op = @stub_method.call request, deadline: deadline, metadata: options.metadata, return_op: true
-
-          res = op.execute
-          yield op if block
-          res
+          operation = stub_method.call request, deadline: deadline, metadata: options.metadata, return_op: true, &block
+
+          if on_stream
+            Thread.new { operation.execute }
+          else
+            response = operation.execute
+            response = on_response.call response if on_response
+            on_operation&.call response, operation
+            response
+          end
         rescue StandardError => error
           if check_retry? deadline
             retry if options.retry_policy.call error
           end
 
           error = Google::Gax.from_error(error).new "RPC failed" if error.is_a? GRPC::BadStatus
 
-          raise error if @exception_transformer.nil?
-          @exception_transformer.call error
```

Comment:
Can we also update https://github.com/googleapis/gapic-generator-ruby/pull/94 with the replacement for this?

Reply:
We don't have a need for it in that PR, as we are only working on the Cloud generator in it. But, the feature is super easy to add. Assume you have a client method like the following:

```ruby
def echo request = nil, options: nil, **request_fields, &block
  # Lots of stuff commented out here...

  @echo ||= Google::Gax::ApiCall.new @echo_stub.method :echo
  @echo.call request, options: options, on_operation: block
end
```

All you need to add is a `rescue` statement, and wrap the raised error:

```ruby
def echo request = nil, options: nil, **request_fields, &block
  # Lots of stuff commented out here...

  @echo ||= Google::Gax::ApiCall.new @echo_stub.method :echo
  @echo.call request, options: options, on_operation: block
rescue Google::Gax::GaxError => error
  raise GoogleAdsError.new error.message
end
```

I really don't see a benefit for swallowing errors in the the Gax layer when the functionality is so easily added using a custom generator (which they are going to be doing anyway).

Reply:
To rephrase: this looks like an Ads-specific feature, and I don't see where it fits into the Cloud generator.

Reply:
Ok, I buy this now. This kind of customization would have been more difficult in the past without the framework but I see your point - now we probably don't need a special knob for something like this.